1. <span style="font-size: 16.000000pt; font-family: 'Cambria'">Which shippers do we have?&nbsp;</span>

In [1]:
SELECT * FROM Shippers

(3 rows affected)

Total execution time: 00:00:00.108

ShipperID CompanyName Phone 1 Speedy Express (503) 555-9831 2 United Package (503) 555-3199 3 Federal Shipping (503) 555-9931

2. <span style="color: var(--vscode-foreground); font-family: Cambria; font-size: 16pt;">Certain fields from Categories&nbsp;</span>

In [2]:
SELECT 
    CategoryName,
    [Description] 
FROM Categories

(8 rows affected)

Total execution time: 00:00:00.010

CategoryName Description Beverages Soft drinks, coffees, teas, beers, and ales Condiments Sweet and savory sauces, relishes, spreads, and seasonings Confections Desserts, candies, and sweet breads Dairy Products Cheeses Grains/Cereals Breads, crackers, pasta, and cereal Meat/Poultry Prepared meats Produce Dried fruit and bean curd Seafood Seaweed and fish

<span style="font-size: 16.000000pt; font-family: 'Cambria'">3. Sales Representatives&nbsp;</span> 

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">We’d like to see just the FirstName, LastName, and HireDate of all the employees with the </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">title of Sales Representative. Write a SQL statement that returns only those employees.&nbsp;</span>

In [3]:
SELECT
    FirstName,
    LastName,
    HireDate
FROM Employees

(9 rows affected)

Total execution time: 00:00:00.009

FirstName LastName HireDate Nancy Davolio 2010-05-01 00:00:00.000 Andrew Fuller 2010-08-14 00:00:00.000 Janet Leverling 2010-04-01 00:00:00.000 Margaret Peacock 2011-05-03 00:00:00.000 Steven Buchanan 2011-10-17 00:00:00.000 Michael Suyama 2011-10-17 00:00:00.000 Robert King 2012-01-02 00:00:00.000 Laura Callahan 2012-03-05 00:00:00.000 Anne Dodsworth 2012-11-15 00:00:00.000

<span style="font-size: 16.000000pt; font-family: 'Cambria'">4. Sales Representatives in the United States&nbsp;</span> 

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Now we’d like to see the same columns as above, but only for those employees that both </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">have the title of Sales Representative, and also are in the United States.&nbsp;</span>

In [4]:
SELECT
    FirstName,
    LastName,
    HireDate
FROM Employees
WHERE Title = 'Sales Representative'
AND Country = 'USA'

(3 rows affected)

Total execution time: 00:00:00.005

FirstName LastName HireDate Nancy Davolio 2010-05-01 00:00:00.000 Janet Leverling 2010-04-01 00:00:00.000 Margaret Peacock 2011-05-03 00:00:00.000

<span style="font-size: 16.000000pt; font-family: 'Cambria'">5. Orders placed by specific EmployeeID</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Show all the orders placed by a specific employee. The EmployeeID for this Employee (Steven Buchanan) is 5.&nbsp;</span>

In [5]:
SELECT 
    OrderID,
    OrderDate
FROM Orders
WHERE EmployeeID = 5

(42 rows affected)

Total execution time: 00:00:00.004

OrderID OrderDate 10248 2014-07-04 08:00:00.000 10254 2014-07-11 02:00:00.000 10269 2014-07-31 00:00:00.000 10297 2014-09-04 21:00:00.000 10320 2014-10-03 12:00:00.000 10333 2014-10-18 18:00:00.000 10358 2014-11-20 05:00:00.000 10359 2014-11-21 14:00:00.000 10372 2014-12-04 10:00:00.000 10378 2014-12-10 00:00:00.000 10397 2014-12-27 17:00:00.000 10463 2015-03-04 13:00:00.000 10474 2015-03-13 16:00:00.000 10477 2015-03-17 02:00:00.000 10529 2015-05-07 01:00:00.000 10549 2015-05-27 03:00:00.000 10569 2015-06-16 15:00:00.000 10575 2015-06-20 22:00:00.000 10607 2015-07-22 09:00:00.000 10648 2015-08-28 22:00:00.000 10649 2015-08-28 00:00:00.000 10650 2015-08-29 06:00:00.000 10654 2015-09-02 07:00:00.000 10675 2015-09-19 06:00:00.000 10711 2015-10-21 03:00:00.000 10714 2015-10-22 03:00:00.000 10721 2015-10-29 08:00:00.000 10730 2015-11-05 07:00:00.000 10761 2015-12-02 08:00:00.000 10812 2016-01-02 02:00:00.000 10823 2016-01-09 17:00:00.000 10841 2016-01-20 21:00:00.000 10851 2016-01-26 00:00:00.000 10866 2016-02-03 01:00:00.000 10869 2016-02-04 09:00:00.000 10870 2016-02-04 12:00:00.000 10872 2016-02-05 06:00:00.000 10874 2016-02-06 14:00:00.000 10899 2016-02-20 09:00:00.000 10922 2016-03-03 02:00:00.000 10954 2016-03-17 16:00:00.000 11043 2016-04-22 17:00:00.000

<span style="font-size: 16.000000pt; font-family: 'Cambria'">6. Suppliers and ContactTitles</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">In the Suppliers table, show the SupplierID, ContactName, and ContactTitle for those Suppliers whose ContactTitle is </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPS'; font-style: italic">not </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Marketing Manager.&nbsp;</span>

In [6]:
SELECT 
    SupplierID,
    ContactName,
    ContactTitle
FROM Suppliers
WHERE ContactTitle != 'Marketing Manager'

(24 rows affected)

Total execution time: 00:00:00.007

SupplierID ContactName ContactTitle 1 Charlotte Cooper Purchasing Manager 2 Shelley Burke Order Administrator 3 Regina Murphy Sales Representative 5 Antonio del Valle Saavedra Export Administrator 6 Mayumi Ohno Marketing Representative 8 Peter Wilson Sales Representative 9 Lars Peterson Sales Agent 11 Petra Winkler Sales Manager 12 Martin Bein International Marketing Mgr. 13 Sven Petersen Coordinator Foreign Markets 14 Elio Rossi Sales Representative 16 Cheryl Saylor Regional Account Rep. 17 Michael Björn Sales Representative 18 Guylène Nodier Sales Manager 19 Robb Merchant Wholesale Account Agent 20 Chandra Leka Owner 21 Niels Petersen Sales Manager 22 Dirk Luchte Accounting Manager 23 Anne Heikkonen Product Manager 24 Wendy Mackenzie Sales Representative 26 Giovanni Giudici Order Administrator 27 Marie Delamare Sales Manager 28 Eliane Noz Sales Representative 29 Chantal Goulet Accounting Manager

<span style="font-size: 16.000000pt; font-family: 'Cambria'">7. Products with </span> <span style="font-size: 16.000000pt; font-family: 'Cambria'">“q</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">ueso</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">” in ProductName</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">In the products table, we’d like to see the ProductID and ProductName for those products </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">where the ProductName includes </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">the string “queso”.&nbsp;</span>

In [7]:
SELECT 
    ProductID,
    ProductName
FROM Products
WHERE ProductName LIKE '%Queso%'


(2 rows affected)

Total execution time: 00:00:00.004

ProductID ProductName 11 Queso Cabrales 12 Queso Manchego La Pastora

<span style="font-size: 16.000000pt; font-family: 'Cambria'">8. Orders shipping to France or Belgium</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Looking at the Orders </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">table, there’s a field called ShipCountry. Write a query that shows </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">the OrderID, CustomerID, and ShipCountry for the orders where the ShipCountry is either France or Belgium.&nbsp;</span>

In [8]:
SELECT 
OrderID,
CustomerID,
ShipCountry
FROM Orders
WHERE ShipCountry in ('France', 'Belgium')

(96 rows affected)

Total execution time: 00:00:00.007

OrderID CustomerID ShipCountry 10248 VINET France 10251 VICTE France 10252 SUPRD Belgium 10265 BLONP France 10274 VINET France 10295 VINET France 10297 BLONP France 10302 SUPRD Belgium 10311 DUMON France 10331 BONAP France 10334 VICTE France 10340 BONAP France 10350 LAMAI France 10358 LAMAI France 10360 BLONP France 10362 BONAP France 10371 LAMAI France 10408 FOLIG France 10413 LAMAI France 10425 LAMAI France 10436 BLONP France 10449 BLONP France 10450 VICTE France 10454 LAMAI France 10458 SUPRD Belgium 10459 VICTE France 10463 SUPRD Belgium 10470 BONAP France 10475 SUPRD Belgium 10478 VICTE France 10480 FOLIG France 10493 LAMAI France 10500 LAMAI France 10511 BONAP France 10525 BONAP France 10529 MAISD Belgium 10546 VICTE France 10559 BLONP France 10566 BLONP France 10584 BLONP France 10609 DUMON France 10610 LAMAI France 10628 BLONP France 10631 LAMAI France 10634 FOLIG France 10649 MAISD Belgium 10663 BONAP France 10671 FRANR France 10679 BLONP France 10683 DUMON France 10715 BONAP France 10730 BONAP France 10732 BONAP France 10737 VINET France 10738 SPECD France 10739 VINET France 10755 BONAP France 10760 MAISD Belgium 10763 FOLIG France 10767 SUPRD Belgium 10787 LAMAI France 10789 FOLIG France 10806 VICTE France 10814 VICTE France 10826 BLONP France 10827 BONAP France 10832 LAMAI France 10841 SUPRD Belgium 10843 VICTE France 10846 SUPRD Belgium 10850 VICTE France 10858 LACOR France 10860 FRANR France 10871 BONAP France 10876 BONAP France 10885 SUPRD Belgium 10890 DUMON France 10892 MAISD Belgium 10896 MAISD Belgium 10907 SPECD France 10923 LAMAI France 10927 LACOR France 10930 SUPRD Belgium 10932 BONAP France 10940 BONAP France 10964 SPECD France 10971 FRANR France 10972 LACOR France 10973 LACOR France 10978 MAISD Belgium 11004 MAISD Belgium 11035 SUPRD Belgium 11038 SUPRD Belgium 11043 SPECD France 11051 LAMAI France 11076 BONAP France

<span style="font-size: 16.000000pt; font-family: 'Cambria'">9. Orders shipping to any country in Latin America</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Now, instead of just wanting to return all the orders from France or Belgium, we want to </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">show all the orders from any Latin American country. But we don’t have a list of Latin American countries in a table in the Northwind database. So, we’re going to just use this </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">list of Latin American countries that happen to be in the Orders table:</span>

In [9]:
SELECT 
OrderID,
CustomerID,
ShipCountry
FROM Orders
WHERE ShipCountry in ('Brazil','Mexico','Argentina','Venezuela')

(173 rows affected)

Total execution time: 00:00:00.005

OrderID CustomerID ShipCountry 10250 HANAR Brazil 10253 HANAR Brazil 10256 WELLI Brazil 10257 HILAA Venezuela 10259 CENTC Mexico 10261 QUEDE Brazil 10268 GROSR Venezuela 10276 TORTU Mexico 10283 LILAS Venezuela 10287 RICAR Brazil 10290 COMMI Brazil 10291 QUEDE Brazil 10292 TRADH Brazil 10293 TORTU Mexico 10296 LILAS Venezuela 10299 RICAR Brazil 10304 TORTU Mexico 10308 ANATR Mexico 10319 TORTU Mexico 10322 PERIC Mexico 10330 LILAS Venezuela 10347 FAMIA Brazil 10354 PERIC Mexico 10357 LILAS Venezuela 10365 ANTON Mexico 10372 QUEEN Brazil 10379 QUEDE Brazil 10381 LILAS Venezuela 10386 FAMIA Brazil 10395 HILAA Venezuela 10405 LINOD Venezuela 10406 QUEEN Brazil 10409 OCEAN Argentina 10414 FAMIA Brazil 10420 WELLI Brazil 10421 QUEDE Brazil 10423 GOURL Brazil 10447 RICAR Brazil 10448 RANCH Argentina 10461 LILAS Venezuela 10466 COMMI Brazil 10474 PERIC Mexico 10476 HILAA Venezuela 10481 RICAR Brazil 10485 LINOD Venezuela 10486 HILAA Venezuela 10487 QUEEN Brazil 10490 HILAA Venezuela 10494 COMMI Brazil 10496 TRADH Brazil 10498 HILAA Venezuela 10499 LILAS Venezuela 10502 PERIC Mexico 10507 ANTON Mexico 10512 FAMIA Brazil 10518 TORTU Mexico 10521 CACTU Argentina 10531 OCEAN Argentina 10535 ANTON Mexico 10541 HANAR Brazil 10543 LILAS Venezuela 10552 HILAA Venezuela 10563 RICAR Brazil 10573 ANTON Mexico 10576 TORTU Mexico 10581 FAMIA Brazil 10585 WELLI Brazil 10587 QUEDE Brazil 10601 HILAA Venezuela 10606 TRADH Brazil 10613 HILAA Venezuela 10622 RICAR Brazil 10625 ANATR Mexico 10637 QUEEN Brazil 10638 LINOD Venezuela 10641 HILAA Venezuela 10644 WELLI Brazil 10645 HANAR Brazil 10647 QUEDE Brazil 10648 RICAR Brazil 10650 FAMIA Brazil 10652 GOURL Brazil 10659 QUEEN Brazil 10676 TORTU Mexico 10677 ANTON Mexico 10682 ANTON Mexico 10685 GOURL Brazil 10690 HANAR Brazil 10697 LINOD Venezuela 10704 QUEEN Brazil 10705 HILAA Venezuela 10709 GOURL Brazil 10716 RANCH Argentina 10720 QUEDE Brazil 10725 FAMIA Brazil 10728 QUEEN Brazil 10729 LINOD Venezuela 10734 GOURL Brazil 10759 ANATR Mexico 10770 HANAR Brazil 10777 GOURL Brazil 10780 LILAS Venezuela 10782 CACTU Argentina 10783 HANAR Brazil 10785 GROSR Venezuela 10786 QUEEN Brazil 10790 GOURL Brazil 10794 QUEDE Brazil 10796 HILAA Venezuela 10803 WELLI Brazil 10809 WELLI Brazil 10811 LINOD Venezuela 10813 RICAR Brazil 10819 CACTU Argentina 10823 LILAS Venezuela 10828 RANCH Argentina 10830 TRADH Brazil 10834 TRADH Brazil 10838 LINOD Venezuela 10839 TRADH Brazil 10840 LINOD Venezuela 10842 TORTU Mexico 10851 RICAR Brazil 10856 ANTON Mexico 10863 HILAA Venezuela 10868 QUEEN Brazil 10877 RICAR Brazil 10881 CACTU Argentina 10886 HANAR Brazil 10898 OCEAN Argentina 10899 LILAS Venezuela 10900 WELLI Brazil 10901 HILAA Venezuela 10903 HANAR Brazil 10905 WELLI Brazil 10913 QUEEN Brazil 10914 QUEEN Brazil 10915 TORTU Mexico 10916 RANCH Argentina 10919 LINOD Venezuela 10922 HANAR Brazil 10925 HANAR Brazil 10926 ANATR Mexico 10935 WELLI Brazil 10937 CACTU Argentina 10954 LINOD Venezuela 10957 HILAA Venezuela 10958 OCEAN Argentina 10959 GOURL Brazil 10960 HILAA Venezuela 10961 QUEEN Brazil 10969 COMMI Brazil 10976 HILAA Venezuela 10981 HANAR Brazil 10986 OCEAN Argentina 10989 QUEDE Brazil 10995 PERIC Mexico 10997 LILAS Venezuela 11014 LINOD Venezuela 11019 RANCH Argentina 11022 HANAR Brazil 11039 LINOD Venezuela 11042 COMMI Brazil 11049 GOURL Brazil 11052 HANAR Brazil 11054 CACTU Argentina 11055 HILAA Venezuela 11059 RICAR Brazil 11065 LILAS Venezuela 11068 QUEEN Brazil 11069 TORTU Mexico 11071 LILAS Venezuela 11073 PERIC Mexico

<span style="font-size: 16.000000pt; font-family: 'Cambria'">10. Employees, in order of age</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">For all the employees in the Employees table, show the FirstName, LastName, Title, and BirthDate. Order the results by BirthDate, so we have the oldest employees first.&nbsp;</span>

In [10]:
SELECT
FirstName, 
LastName,
Title,
BirthDate
FROM Employees
ORDER BY BirthDate

(9 rows affected)

Total execution time: 00:00:00.011

FirstName LastName Title BirthDate Margaret Peacock Sales Representative 1955-09-19 00:00:00.000 Nancy Davolio Sales Representative 1966-12-08 00:00:00.000 Andrew Fuller Vice President, Sales 1970-02-19 00:00:00.000 Steven Buchanan Sales Manager 1973-03-04 00:00:00.000 Laura Callahan Inside Sales Coordinator 1976-01-09 00:00:00.000 Robert King Sales Representative 1978-05-29 00:00:00.000 Michael Suyama Sales Representative 1981-07-02 00:00:00.000 Janet Leverling Sales Representative 1981-08-30 00:00:00.000 Anne Dodsworth Sales Representative 1984-01-27 00:00:00.000

<span style="font-size: 16.000000pt; font-family: 'Cambria'">11. Showing only the Date with a DateTime field</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">In the output of the query above, showing the Employees in order of BirthDate, we see the time of the BirthDate </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">field, which we don’t </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">want. Show only the date portion of the BirthDate field.&nbsp;</span>

In [11]:
SELECT
    FirstName,
    LastName,
    Title,
    DateOnlyBirthDate = CONVERT(date, BirthDate)
FROM Employees

(9 rows affected)

Total execution time: 00:00:00.004

FirstName LastName Title DateOnlyBirthDate Nancy Davolio Sales Representative 1966-12-08 Andrew Fuller Vice President, Sales 1970-02-19 Janet Leverling Sales Representative 1981-08-30 Margaret Peacock Sales Representative 1955-09-19 Steven Buchanan Sales Manager 1973-03-04 Michael Suyama Sales Representative 1981-07-02 Robert King Sales Representative 1978-05-29 Laura Callahan Inside Sales Coordinator 1976-01-09 Anne Dodsworth Sales Representative 1984-01-27

<span style="font-size: 16.000000pt; font-family: 'Cambria'">12. Employees full name</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Show the FirstName and LastName columns from the Employees table, and then create a new column called FullName, showing FirstName and LastName joined together in one column, with a space in-between.&nbsp;</span>

In [12]:
SELECT
    FirstName,
    LastName,
    FullName = CONCAT(FirstName, ' ', LastName)
FROM Employees

(9 rows affected)

Total execution time: 00:00:00.003

FirstName LastName FullName Nancy Davolio Nancy Davolio Andrew Fuller Andrew Fuller Janet Leverling Janet Leverling Margaret Peacock Margaret Peacock Steven Buchanan Steven Buchanan Michael Suyama Michael Suyama Robert King Robert King Laura Callahan Laura Callahan Anne Dodsworth Anne Dodsworth

<span style="font-size: 16.000000pt; font-family: 'Cambria'">13. OrderDetails amount per line item</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">In the OrderDetails table, we have the fields UnitPrice and Quantity. Create a new field, TotalPrice, that multiplies these two together. </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">We’ll ignore the Discount field for now.</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">In addition, show the OrderID, ProductID, UnitPrice, and Quantity. Order by OrderID and ProductID.&nbsp;</span>

In [13]:
SELECT 
    OrderID,
    ProductID,
    UnitPrice,
    Quantity,
    TotalPrice = CONCAT('$ ', UnitPrice * Quantity)
FROM OrderDetails

(2155 rows affected)

Total execution time: 00:00:00.033

OrderID ProductID UnitPrice Quantity TotalPrice 10248 11 14.00 12 $ 168.00 10248 42 9.80 10 $ 98.00 10248 72 34.80 5 $ 174.00 10249 14 18.60 9 $ 167.40 10249 51 42.40 40 $ 1696.00 10250 41 7.70 10 $ 77.00 10250 51 42.40 35 $ 1484.00 10250 65 16.80 15 $ 252.00 10251 22 16.80 6 $ 100.80 10251 57 15.60 15 $ 234.00 10251 65 16.80 20 $ 336.00 10252 20 64.80 40 $ 2592.00 10252 33 2.00 25 $ 50.00 10252 60 27.20 40 $ 1088.00 10253 31 10.00 20 $ 200.00 10253 39 14.40 42 $ 604.80 10253 49 16.00 40 $ 640.00 10254 24 3.60 15 $ 54.00 10254 55 19.20 21 $ 403.20 10254 74 8.00 21 $ 168.00 10255 2 15.20 20 $ 304.00 10255 16 13.90 35 $ 486.50 10255 36 15.20 25 $ 380.00 10255 59 44.00 30 $ 1320.00 10256 53 26.20 15 $ 393.00 10256 77 10.40 12 $ 124.80 10257 27 35.10 25 $ 877.50 10257 39 14.40 6 $ 86.40 10257 77 10.40 15 $ 156.00 10258 2 15.20 50 $ 760.00 10258 5 17.00 65 $ 1105.00 10258 32 25.60 6 $ 153.60 10259 21 8.00 10 $ 80.00 10259 37 20.80 1 $ 20.80 10260 41 7.70 16 $ 123.20 10260 57 15.60 50 $ 780.00 10260 62 39.40 15 $ 591.00 10260 70 12.00 21 $ 252.00 10261 21 8.00 20 $ 160.00 10261 35 14.40 20 $ 288.00 10262 5 17.00 12 $ 204.00 10262 7 24.00 15 $ 360.00 10262 56 30.40 2 $ 60.80 10263 16 13.90 60 $ 834.00 10263 24 3.60 65 $ 234.00 10263 30 20.70 60 $ 1242.00 10263 74 8.00 65 $ 520.00 10264 2 15.20 35 $ 532.00 10264 41 7.70 25 $ 192.50 10265 17 31.20 30 $ 936.00 10265 70 12.00 20 $ 240.00 10266 12 30.40 12 $ 364.80 10267 40 14.70 50 $ 735.00 10267 59 44.00 70 $ 3080.00 10267 76 14.40 15 $ 216.00 10268 29 99.00 10 $ 990.00 10268 72 27.80 4 $ 111.20 10269 33 2.00 60 $ 120.00 10269 72 27.80 20 $ 556.00 10270 36 15.20 30 $ 456.00 10270 43 36.80 25 $ 920.00 10271 33 2.00 24 $ 48.00 10272 20 64.80 6 $ 388.80 10272 31 10.00 40 $ 400.00 10272 72 27.80 24 $ 667.20 10273 10 24.80 24 $ 595.20 10273 31 10.00 15 $ 150.00 10273 33 2.00 20 $ 40.00 10273 40 14.70 60 $ 882.00 10273 76 14.40 33 $ 475.20 10274 71 17.20 20 $ 344.00 10274 72 27.80 7 $ 194.60 10275 24 3.60 12 $ 43.20 10275 59 44.00 6 $ 264.00 10276 10 24.80 15 $ 372.00 10276 13 4.80 10 $ 48.00 10277 28 36.40 20 $ 728.00 10277 62 39.40 12 $ 472.80 10278 44 15.50 16 $ 248.00 10278 59 44.00 15 $ 660.00 10278 63 35.10 8 $ 280.80 10278 73 12.00 25 $ 300.00 10279 17 31.20 15 $ 468.00 10280 24 3.60 12 $ 43.20 10280 55 19.20 20 $ 384.00 10280 75 6.20 30 $ 186.00 10281 19 7.30 1 $ 7.30 10281 24 3.60 6 $ 21.60 10281 35 14.40 4 $ 57.60 10282 30 20.70 6 $ 124.20 10282 57 15.60 2 $ 31.20 10283 15 12.40 20 $ 248.00 10283 19 7.30 18 $ 131.40 10283 60 27.20 35 $ 952.00 10283 72 27.80 3 $ 83.40 10284 27 35.10 15 $ 526.50 10284 44 15.50 21 $ 325.50 10284 60 27.20 20 $ 544.00 10284 67 11.20 5 $ 56.00 10285 1 14.40 45 $ 648.00 10285 40 14.70 40 $ 588.00 10285 53 26.20 36 $ 943.20 10286 35 14.40 100 $ 1440.00 10286 62 39.40 40 $ 1576.00 10287 16 13.90 40 $ 556.00 10287 34 11.20 20 $ 224.00 10287 46 9.60 15 $ 144.00 10288 54 5.90 10 $ 59.00 10288 68 10.00 3 $ 30.00 10289 3 8.00 30 $ 240.00 10289 64 26.60 9 $ 239.40 10290 5 17.00 20 $ 340.00 10290 29 99.00 15 $ 1485.00 10290 49 16.00 15 $ 240.00 10290 77 10.40 10 $ 104.00 10291 13 4.80 20 $ 96.00 10291 44 15.50 24 $ 372.00 10291 51 42.40 2 $ 84.80 10292 20 64.80 20 $ 1296.00 10293 18 50.00 12 $ 600.00 10293 24 3.60 10 $ 36.00 10293 63 35.10 5 $ 175.50 10293 75 6.20 6 $ 37.20 10294 1 14.40 18 $ 259.20 10294 17 31.20 15 $ 468.00 10294 43 36.80 15 $ 552.00 10294 60 27.20 21 $ 571.20 10294 75 6.20 6 $ 37.20 10295 56 30.40 4 $ 121.60 10296 11 16.80 12 $ 201.60 10296 16 13.90 30 $ 417.00 10296 69 28.80 15 $ 432.00 10297 39 14.40 60 $ 864.00 10297 72 27.80 20 $ 556.00 10298 2 15.20 40 $ 608.00 10298 36 15.20 40 $ 608.00 10298 59 44.00 30 $ 1320.00 10298 62 39.40 15 $ 591.00 10299 19 7.30 15 $ 109.50 10299 70 12.00 20 $ 240.00 10300 66 13.60 30 $ 408.00 10300 68 10.00 20 $ 200.00 10301 40 14.70 10 $ 147.00 10301 56 30.40 20 $ 608.00 10302 17 31.20 40 $ 1248.00 10302 28 36.40 28 $ 1019.20 10302 43 36.80 12 $ 441.60 10303 40 14.70 40 $ 588.00 10303 65 16.80 30 $ 504.00 10303 

<span style="font-size: 16.000000pt; font-family: 'Cambria'">14. How many customers?</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">How many customers do we have in the Customers table? Show one value only, and </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">don’t </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">rely on getting the record count at the end of a resultset.</span>

In [14]:
SELECT 
    TotalCustomers = COUNT(*)
FROM Customers

(1 row affected)

Total execution time: 00:00:00.004

TotalCustomers 91

<span style="font-size: 16.000000pt; font-family: 'Cambria'">15. When was the first order?</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Show the date of the first order ever made in the Orders table.&nbsp;</span>

In [15]:
SELECT TOP 1
    FirstOrder = OrderDate
FROM Orders
ORDER BY OrderDate

(1 row affected)

Total execution time: 00:00:00.003

FirstOrder 2014-07-04 08:00:00.000

In [16]:
SELECT
    FirstOrder = MIN(OrderDate)
FROM Orders

(1 row affected)

Total execution time: 00:00:00.003

FirstOrder 2014-07-04 08:00:00.000

<span style="font-size: 16.000000pt; font-family: 'Cambria'">16. Countries where there are customers</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Show a list of countries where the Northwind company has customers.&nbsp;</span>

In [17]:
SELECT 
    Country
FROM Customers
GROUP BY Country



(21 rows affected)

Total execution time: 00:00:00.006

Country Argentina Austria Belgium Brazil Canada Denmark Finland France Germany Ireland Italy Mexico Norway Poland Portugal Spain Sweden Switzerland UK USA Venezuela

<span style="font-size: 16.000000pt; font-family: 'Cambria'">17. Contact titles for customers</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Show a list of all the different values in the Customers table for ContactTitles. Also include a count for each ContactTitle.</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">This is similar in concept to the previous question “</span><span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Countries where there are customers</span><span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">”, </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">except we now want a count for each ContactTitle.&nbsp;</span>

In [18]:
SELECT 
    ContactTitle,
    TotalContactTitle = COUNT(*)
FROM Customers
GROUP BY ContactTitle
ORDER BY TotalContactTitle DESC

(12 rows affected)

Total execution time: 00:00:00.004

ContactTitle TotalContactTitle Owner 17 Sales Representative 17 Marketing Manager 12 Sales Manager 11 Accounting Manager 10 Sales Associate 7 Marketing Assistant 6 Sales Agent 5 Assistant Sales Agent 2 Order Administrator 2 Assistant Sales Representative 1 Owner/Marketing Assistant 1

<span style="font-size: 16.000000pt; font-family: 'Cambria'">18. Products with associated supplier names</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">We’d like to show, for each product, the associated </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Supplier. Show the ProductID, ProductName, and the CompanyName of the Supplier.</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Sort the result by ProductID.</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">This question will introduce what may be a new concept</span><span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">—</span><span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">the Join clause in SQL. The Join clause is used to join two or more relational database tables together in a logical way. In this case, you will join the Suppliers table to the Products table by SupplierID.&nbsp;</span>

In [19]:
SELECT 
    P.ProductID,
    P.ProductName,
    Supplier = S.CompanyName
FROM Products P
JOIN Suppliers S ON P.SupplierID = S.SupplierID 


(77 rows affected)

Total execution time: 00:00:00.006

ProductID ProductName Supplier 1 Chai Exotic Liquids 2 Chang Exotic Liquids 3 Aniseed Syrup Exotic Liquids 4 Chef Anton's Cajun Seasoning New Orleans Cajun Delights 5 Chef Anton's Gumbo Mix New Orleans Cajun Delights 6 Grandma's Boysenberry Spread Grandma Kelly's Homestead 7 Uncle Bob's Organic Dried Pears Grandma Kelly's Homestead 8 Northwoods Cranberry Sauce Grandma Kelly's Homestead 9 Mishi Kobe Niku Tokyo Traders 10 Ikura Tokyo Traders 11 Queso Cabrales Cooperativa de Quesos 'Las Cabras' 12 Queso Manchego La Pastora Cooperativa de Quesos 'Las Cabras' 13 Konbu Mayumi's 14 Tofu Mayumi's 15 Genen Shouyu Mayumi's 16 Pavlova Pavlova, Ltd. 17 Alice Mutton Pavlova, Ltd. 18 Carnarvon Tigers Pavlova, Ltd. 19 Teatime Chocolate Biscuits Specialty Biscuits, Ltd. 20 Sir Rodney's Marmalade Specialty Biscuits, Ltd. 21 Sir Rodney's Scones Specialty Biscuits, Ltd. 22 Gustaf's Knäckebröd PB Knäckebröd AB 23 Tunnbröd PB Knäckebröd AB 24 Guaraná Fantástica Refrescos Americanas LTDA 25 NuNuCa Nuß-Nougat-Creme Heli Süßwaren GmbH & Co. KG 26 Gumbär Gummibärchen Heli Süßwaren GmbH & Co. KG 27 Schoggi Schokolade Heli Süßwaren GmbH & Co. KG 28 Rössle Sauerkraut Plutzer Lebensmittelgroßmärkte AG 29 Thüringer Rostbratwurst Plutzer Lebensmittelgroßmärkte AG 30 Nord-Ost Matjeshering Nord-Ost-Fisch Handelsgesellschaft mbH 31 Gorgonzola Telino Formaggi Fortini s.r.l. 32 Mascarpone Fabioli Formaggi Fortini s.r.l. 33 Geitost Norske Meierier 34 Sasquatch Ale Bigfoot Breweries 35 Steeleye Stout Bigfoot Breweries 36 Inlagd Sill Svensk Sjöföda AB 37 Gravad lax Svensk Sjöföda AB 38 Côte de Blaye Aux joyeux ecclésiastiques 39 Chartreuse verte Aux joyeux ecclésiastiques 40 Boston Crab Meat New England Seafood Cannery 41 Jack's New England Clam Chowder New England Seafood Cannery 42 Singaporean Hokkien Fried Mee Leka Trading 43 Ipoh Coffee Leka Trading 44 Gula Malacca Leka Trading 45 Rogede sild Lyngbysild 46 Spegesild Lyngbysild 47 Zaanse koeken Zaanse Snoepfabriek 48 Chocolade Zaanse Snoepfabriek 49 Maxilaku Karkki Oy 50 Valkoinen suklaa Karkki Oy 51 Manjimup Dried Apples G'day, Mate 52 Filo Mix G'day, Mate 53 Perth Pasties G'day, Mate 54 Tourtière Ma Maison 55 Pâté chinois Ma Maison 56 Gnocchi di nonna Alice Pasta Buttini s.r.l. 57 Ravioli Angelo Pasta Buttini s.r.l. 58 Escargots de Bourgogne Escargots Nouveaux 59 Raclette Courdavault Gai pâturage 60 Camembert Pierrot Gai pâturage 61 Sirop d'érable Forêts d'érables 62 Tarte au sucre Forêts d'érables 63 Vegie-spread Pavlova, Ltd. 64 Wimmers gute Semmelknödel Plutzer Lebensmittelgroßmärkte AG 65 Louisiana Fiery Hot Pepper Sauce New Orleans Cajun Delights 66 Louisiana Hot Spiced Okra New Orleans Cajun Delights 67 Laughing Lumberjack Lager Bigfoot Breweries 68 Scottish Longbreads Specialty Biscuits, Ltd. 69 Gudbrandsdalsost Norske Meierier 70 Outback Lager Pavlova, Ltd. 71 Flotemysost Norske Meierier 72 Mozzarella di Giovanni Formaggi Fortini s.r.l. 73 Röd Kaviar Svensk Sjöföda AB 74 Longlife Tofu Tokyo Traders 75 Rhönbräu Klosterbier Plutzer Lebensmittelgroßmärkte AG 76 Lakkalikööri Karkki Oy 77 Original Frankfurter grüne Soße Plutzer Lebensmittelgroßmärkte AG

<span style="font-size: 16.000000pt; font-family: 'Cambria'">19. Orders and the Shipper that was used</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">We’d like to show a list of the Orders that were made, including the </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Shipper that was used. Show the OrderID, OrderDate (date only), and CompanyName of the Shipper, and sort by OrderID.</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">In order to not show all the orders (</span><span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">there’s more than 800</span><span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">), show only those rows with an OrderID of less than </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'; background-color: rgb(100.000000%, 100.000000%, 100.000000%)">10270</span><span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">.&nbsp;</span>

In [20]:
SELECT 
    O.OrderID,
    OrderDate = CONVERT(date, O.OrderDate),
    Shipper = S.CompanyName
FROM Orders O 
JOIN Shippers S ON O.ShipVia = S.ShipperID

(830 rows affected)

Total execution time: 00:00:00.011

OrderID OrderDate Shipper 10248 2014-07-04 Federal Shipping 10249 2014-07-05 Speedy Express 10250 2014-07-08 United Package 10251 2014-07-08 Speedy Express 10252 2014-07-09 United Package 10253 2014-07-10 United Package 10254 2014-07-11 United Package 10255 2014-07-12 Federal Shipping 10256 2014-07-15 United Package 10257 2014-07-16 Federal Shipping 10258 2014-07-17 Speedy Express 10259 2014-07-18 Federal Shipping 10260 2014-07-19 Speedy Express 10261 2014-07-19 United Package 10262 2014-07-22 Federal Shipping 10263 2014-07-23 Federal Shipping 10264 2014-07-24 Federal Shipping 10265 2014-07-25 Speedy Express 10266 2014-07-26 Federal Shipping 10267 2014-07-29 Speedy Express 10268 2014-07-30 Federal Shipping 10269 2014-07-31 Speedy Express 10270 2014-08-01 Speedy Express 10271 2014-08-01 United Package 10272 2014-08-02 United Package 10273 2014-08-05 Federal Shipping 10274 2014-08-06 Speedy Express 10275 2014-08-07 Speedy Express 10276 2014-08-08 Federal Shipping 10277 2014-08-09 Federal Shipping 10278 2014-08-12 United Package 10279 2014-08-13 United Package 10280 2014-08-14 Speedy Express 10281 2014-08-14 Speedy Express 10282 2014-08-15 Speedy Express 10283 2014-08-16 Federal Shipping 10284 2014-08-19 Speedy Express 10285 2014-08-20 United Package 10286 2014-08-21 Federal Shipping 10287 2014-08-22 Federal Shipping 10288 2014-08-23 Speedy Express 10289 2014-08-26 Federal Shipping 10290 2014-08-27 Speedy Express 10291 2014-08-27 United Package 10292 2014-08-28 United Package 10293 2014-08-29 Federal Shipping 10294 2014-08-30 United Package 10295 2014-09-02 United Package 10296 2014-09-03 Speedy Express 10297 2014-09-04 United Package 10298 2014-09-05 United Package 10299 2014-09-06 United Package 10300 2014-09-09 United Package 10301 2014-09-09 United Package 10302 2014-09-10 United Package 10303 2014-09-11 United Package 10304 2014-09-12 United Package 10305 2014-09-13 Federal Shipping 10306 2014-09-16 Federal Shipping 10307 2014-09-17 United Package 10308 2014-09-18 Federal Shipping 10309 2014-09-19 Speedy Express 10310 2014-09-20 United Package 10311 2014-09-20 Federal Shipping 10312 2014-09-23 United Package 10313 2014-09-24 United Package 10314 2014-09-25 United Package 10315 2014-09-26 United Package 10316 2014-09-27 Federal Shipping 10317 2014-09-30 Speedy Express 10318 2014-10-01 United Package 10319 2014-10-02 Federal Shipping 10320 2014-10-03 Federal Shipping 10321 2014-10-03 United Package 10322 2014-10-04 Federal Shipping 10323 2014-10-07 Speedy Express 10324 2014-10-08 Speedy Express 10325 2014-10-09 Federal Shipping 10326 2014-10-10 United Package 10327 2014-10-11 Speedy Express 10328 2014-10-14 Federal Shipping 10329 2014-10-15 United Package 10330 2014-10-16 Speedy Express 10331 2014-10-16 Speedy Express 10332 2014-10-17 United Package 10333 2014-10-18 Federal Shipping 10334 2014-10-21 United Package 10335 2014-10-22 United Package 10336 2014-10-23 United Package 10337 2014-10-24 Federal Shipping 10338 2014-10-25 Federal Shipping 10339 2014-10-28 United Package 10340 2014-10-29 Federal Shipping 10341 2014-10-29 Federal Shipping 10342 2014-10-30 United Package 10343 2014-10-31 Speedy Express 10344 2014-11-01 United Package 10345 2014-11-04 United Package 10346 2014-11-05 Federal Shipping 10347 2014-11-06 Federal Shipping 10348 2014-11-07 United Package 10349 2014-11-08 Speedy Express 10350 2014-11-11 United Package 10351 2014-11-11 Speedy Express 10352 2014-11-12 Federal Shipping 10353 2014-11-13 Federal Shipping 10354 2014-11-14 Federal Shipping 10355 2014-11-15 Speedy Express 10356 2014-11-18 United Package 10357 2014-11-19 Federal Shipping 10358 2014-11-20 Speedy Express 10359 2014-11-21 Federal Shipping 10360 2014-11-22 Federal Shipping 10361 2014-11-22 United Package 10362 2014-11-25 Speedy Express 10363 2014-11-26 Federal Shipping 10364 2014-11-26 Speedy Express 10365 2014-11-27 United Package 10366 2014-11-28 United Package 10367 2014-11-28 Federal Shipping 10368 2014-11-29 United Package 10369 2014-12-02 U

<span style="font-size: 16.000000pt; font-family: 'Cambria'">20. Categories, and the total products in each category</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">For this problem, we’d like to see the total number </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">of products in each category. Sort the results by the total number of products, in descending order.&nbsp;</span>

In [21]:
SELECT 
    C.CategoryName,
    TotalProducts = COUNT(*)
FROM Products P 
JOIN Categories C ON P.CategoryID = C.CategoryID
GROUP BY C.CategoryName
ORDER BY TotalProducts DESC

(8 rows affected)

Total execution time: 00:00:00.003

CategoryName TotalProducts Confections 13 Beverages 12 Condiments 12 Seafood 12 Dairy Products 10 Grains/Cereals 7 Meat/Poultry 6 Produce 5

<span style="font-size: 16.000000pt; font-family: 'Cambria'">21. Total customers per country/city</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">In the Customers table, show the total number of customers per Country and City.&nbsp;</span>

In [22]:
SELECT 
    Country,
    City,
    TotalCustomers = COUNT(*)
FROM Customers
GROUP BY Country,City
ORDER BY TotalCustomers DESC

(69 rows affected)

Total execution time: 00:00:00.006

Country City TotalCustomers UK London 6 Mexico México D.F. 5 Brazil Sao Paulo 4 Brazil Rio de Janeiro 3 Spain Madrid 3 Argentina Buenos Aires 3 France Paris 2 USA Portland 2 France Nantes 2 Portugal Lisboa 2 Finland Oulu 1 Italy Reggio Emilia 1 France Reims 1 Brazil Resende 1 Austria Salzburg 1 Venezuela San Cristóbal 1 USA San Francisco 1 USA Seattle 1 Spain Sevilla 1 Norway Stavern 1 France Strasbourg 1 Germany Stuttgart 1 Italy Torino 1 France Toulouse 1 Canada Tsawassen 1 Canada Vancouver 1 France Versailles 1 USA Walla Walla 1 Poland Warszawa 1 USA Butte 1 Brazil Campinas 1 Venezuela Caracas 1 Belgium Charleroi 1 Ireland Cork 1 UK Cowes 1 Germany Cunewalde 1 USA Elgin 1 USA Eugene 1 Germany Frankfurt a.M. 1 Switzerland Genève 1 Austria Graz 1 Finland Helsinki 1 Venezuela I. de Margarita 1 USA Kirkland 1 Denmark Kobenhavn 1 Germany Köln 1 USA Lander 1 Germany Leipzig 1 France Lille 1 Germany Mannheim 1 France Marseille 1 Sweden Luleå 1 France Lyon 1 Canada Montréal 1 Germany München 1 Germany Münster 1 Germany Aachen 1 USA Albuquerque 1 USA Anchorage 1 Denmark Århus 1 Spain Barcelona 1 Venezuela Barquisimeto 1 Italy Bergamo 1 Germany Berlin 1 Switzerland Bern 1 USA Boise 1 Sweden Bräcke 1 Germany Brandenburg 1 Belgium Bruxelles 1

<span style="font-size: 16.000000pt; font-family: 'Cambria'">22. Products that need reordering</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">What products do we have in our inventory that should be reordered? For now, just use the fields UnitsInStock and ReorderLevel, where UnitsInStock is less than or equal to the ReorderLevel, Ignore the fields UnitsOnOrder and Discontinued.</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Sort the results by ProductID.&nbsp;</span>

In [23]:
SELECT 
ProductID,
ProductName,
UnitsInStock,
ReorderLevel
FROM Products
WHERE UnitsInStock <= ReorderLevel
ORDER BY ProductID

(22 rows affected)

Total execution time: 00:00:00.004

ProductID ProductName UnitsInStock ReorderLevel 2 Chang 17 25 3 Aniseed Syrup 13 25 5 Chef Anton's Gumbo Mix 0 0 11 Queso Cabrales 22 30 17 Alice Mutton 0 0 21 Sir Rodney's Scones 3 5 29 Thüringer Rostbratwurst 0 0 30 Nord-Ost Matjeshering 10 15 31 Gorgonzola Telino 0 20 32 Mascarpone Fabioli 9 25 37 Gravad lax 11 25 43 Ipoh Coffee 17 25 45 Rogede sild 5 15 48 Chocolade 15 25 49 Maxilaku 10 15 53 Perth Pasties 0 0 56 Gnocchi di nonna Alice 21 30 64 Wimmers gute Semmelknödel 22 30 66 Louisiana Hot Spiced Okra 4 20 68 Scottish Longbreads 6 15 70 Outback Lager 15 30 74 Longlife Tofu 4 5

<span style="font-size: 16.000000pt; font-family: 'Cambria'">23. Products that need reordering, continued</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Now we need to incorporate these fields</span><span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">—</span><span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">UnitsInStock, UnitsOnOrder, ReorderLevel, Discontinued</span><span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">—into our calculation. We’ll define “products that need reordering” with the </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">following:</span>

- <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">UnitsInStock plus UnitsOnOrder are less than or equal to ReorderLevel</span>
    
- <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">The Discontinued flag is false (0).&nbsp;</span>

In [24]:
SELECT 
ProductID,
ProductName,
UnitsInStock,
UnitsOnOrder,
ReorderLevel,
Discontinued
FROM Products
WHERE UnitsInStock+UnitsOnOrder <= ReorderLevel
AND Discontinued = 0
ORDER BY ProductID

(2 rows affected)

Total execution time: 00:00:00.005

ProductID ProductName UnitsInStock UnitsOnOrder ReorderLevel Discontinued 30 Nord-Ost Matjeshering 10 0 15 0 70 Outback Lager 15 10 30 0

<span style="font-size: 16.000000pt; font-family: 'Cambria'">24.</span><span style="color: var(--vscode-foreground); font-family: Cambria; font-size: 16pt;">Customer list by region</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">A salesperson for Northwind is going on a business trip to visit customers, and would like to see a list of all customers, sorted by region, alphabetically.&nbsp;</span>

In [25]:
SELECT 
    CustomerID,
    CompanyName,
    Region
FROM Customers
ORDER BY 
    CASE 
        WHEN Region IS NULL THEN 1
        ELSE 0
    END,
Region, CustomerID

(91 rows affected)

Total execution time: 00:00:00.005

CustomerID CompanyName Region OLDWO Old World Delicatessen AK BOTTM Bottom-Dollar Markets BC LAUGB Laughing Bacchus Wine Cellars BC LETSS Let's Stop N Shop CA HUNGO Hungry Owl All-Night Grocers Co. Cork GROSR GROSELLA-Restaurante DF SAVEA Save-a-lot Markets ID ISLAT Island Trading Isle of Wight LILAS LILA-Supermercado Lara THECR The Cracker Box MT RATTC Rattlesnake Canyon Grocery NM LINOD LINO-Delicateses Nueva Esparta GREAL Great Lakes Food Market OR HUNGC Hungry Coyote Import Store OR LONEP Lonesome Pine Restaurant OR THEBI The Big Cheese OR MEREP Mère Paillarde Québec HANAR Hanari Carnes RJ QUEDE Que Delícia RJ RICAR Ricardo Adocicados RJ COMMI Comércio Mineiro SP FAMIA Familia Arquibaldo SP GOURL Gourmet Lanchonetes SP QUEEN Queen Cozinha SP TRADH Tradição Hipermercados SP WELLI Wellington Importadora SP HILAA HILARION-Abastos Táchira LAZYK Lazy K Kountry Store WA TRAIH Trail's Head Gourmet Provisioners WA WHITC White Clover Markets WA SPLIR Split Rail Beer & Ale WY ALFKI Alfreds Futterkiste NULL ANATR Ana Trujillo Emparedados y helados NULL ANTON Antonio Moreno Taquería NULL AROUT Around the Horn NULL BERGS Berglunds snabbköp NULL BLAUS Blauer See Delikatessen NULL BLONP Blondesddsl père et fils NULL BOLID Bólido Comidas preparadas NULL BONAP Bon app' NULL BSBEV B's Beverages NULL CACTU Cactus Comidas para llevar NULL CENTC Centro comercial Moctezuma NULL CHOPS Chop-suey Chinese NULL CONSH Consolidated Holdings NULL DRACD Drachenblut Delikatessen NULL DUMON Du monde entier NULL EASTC Eastern Connection NULL ERNSH Ernst Handel NULL FISSA FISSA Fabrica Inter. Salchichas S.A. NULL FOLIG Folies gourmandes NULL FOLKO Folk och fä HB NULL FRANK Frankenversand NULL FRANR France restauration NULL FRANS Franchi S.p.A. NULL FURIB Furia Bacalhau e Frutos do Mar NULL GALED Galería del gastrónomo NULL GODOS Godos Cocina Típica NULL KOENE Königlich Essen NULL LACOR La corne d'abondance NULL LAMAI La maison d'Asie NULL LEHMS Lehmanns Marktstand NULL MAGAA Magazzini Alimentari Riuniti NULL MAISD Maison Dewey NULL MORGK Morgenstern Gesundkost NULL NORTS North/South NULL OCEAN Océano Atlántico Ltda. NULL OTTIK Ottilies Käseladen NULL PARIS Paris spécialités NULL PERIC Pericles Comidas clásicas NULL PICCO Piccolo und mehr NULL PRINI Princesa Isabel Vinhos NULL QUICK QUICK-Stop NULL RANCH Rancho grande NULL REGGC Reggiani Caseifici NULL RICSU Richter Supermarkt NULL ROMEY Romero y tomillo NULL SANTG Santé Gourmet NULL SEVES Seven Seas Imports NULL SIMOB Simons bistro NULL SPECD Spécialités du monde NULL SUPRD Suprêmes délices NULL TOMSP Toms Spezialitäten NULL TORTU Tortuga Restaurante NULL VAFFE Vaffeljernet NULL VICTE Victuailles en stock NULL VINET Vins et alcools Chevalier NULL WANDK Die Wandernde Kuh NULL WARTH Wartian Herkku NULL WILMK Wilman Kala NULL WOLZA Wolski Zajazd NULL

<span style="font-size: 16.000000pt; font-family: 'Cambria'">25. High freight charges</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Some of the countries we ship to have very high freight charges. We'd like to investigate some more shipping options for our customers, to be able to offer them lower freight charges. Return the three ship countries with the highest average freight overall, in descending order by average freight.&nbsp;</span>

In [26]:
SELECT TOP 3
    ShipCountry,
    AverageFreight = AVG(Freight)
FROM Orders
GROUP BY ShipCountry
ORDER BY AverageFreight DESC

(3 rows affected)

Total execution time: 00:00:00.005

ShipCountry AverageFreight Austria 184.7875 Ireland 145.0126 USA 112.8794

<span style="font-size: 16.000000pt; font-family: 'Cambria'">26. High freight charges</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">—</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">2015</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">We're continuing on the question above on high freight charges. Now, instead of using </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPS'; font-style: italic">all </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">the orders we have, we only want to see orders from the year 2015.&nbsp;</span>

In [27]:
SELECT TOP 3
    ShipCountry,
    AverageFreight = AVG(Freight)
FROM Orders
WHERE YEAR(OrderDate) = 2015
GROUP BY ShipCountry
ORDER BY AverageFreight DESC

(3 rows affected)

Total execution time: 00:00:00.003

ShipCountry AverageFreight Austria 178.3642 Switzerland 117.1775 France 113.991

<span style="font-size: 16.000000pt; font-family: 'Cambria'">27. High freight charges with between</span>

In [28]:
Select Top 3
ShipCountry
,AverageFreight = avg(freight)
From Orders
Where
OrderDate between '2014-12-31' and '2016-01-01'
Group By ShipCountry
Order By AverageFreight desc

(3 rows affected)

Total execution time: 00:00:00.004

ShipCountry AverageFreight Austria 178.3642 Switzerland 117.1775 France 113.991

<span style="font-size: 16.000000pt; font-family: 'Cambria'">28. High freight charges</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">—</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">last year</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">We're continuing to work on high freight charges. We now want to get the three ship countries with the highest average freight charges. But instead of filtering for a particular year, we want to use the last 12 months of order data, using as the end date the last OrderDate in Orders.&nbsp;</span>

In [29]:
WITH Mostrecent AS
    (SELECT
        Maxorder = MAX(OrderDate) 
     FROM Orders   
        )

SELECT TOP 3
    ShipCountry,
    AverageFreight = AVG(freight)
FROM Orders
WHERE OrderDate >= DATEADD(yy, -1,(SELECT * FROM Mostrecent))
GROUP BY ShipCountry
ORDER BY AverageFreight DESC


(3 rows affected)

Total execution time: 00:00:00.006

ShipCountry AverageFreight Ireland 200.21 Austria 186.4596 USA 119.3032

In [30]:
SELECT TOP 3
    ShipCountry,
    AverageFreight = AVG(Freight)
FROM Orders
WHERE OrderDate >= DATEADD(yy, -1, (SELECT MAX(OrderDate)FROM Orders))
GROUP BY ShipCountry
ORDER BY AverageFreight DESC

(3 rows affected)

Total execution time: 00:00:00.005

ShipCountry AverageFreight Ireland 200.21 Austria 186.4596 USA 119.3032

<span style="font-size: 16.000000pt; font-family: 'Cambria'">29. Employee/Order Detail report</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">We're doing inventory, and need to show Employee and Order Detail information like the below, for all orders. Sort by OrderID and Product ID.&nbsp;</span>

In [31]:
SELECT 
    O.EmployeeID,
    E.LastName,
    OD.OrderID,
    P.ProductName,
    OD.Quantity
FROM OrderDetails OD 
JOIN Orders O ON OD.OrderID = O.OrderID
JOIN Employees E ON O.EmployeeID = E.EmployeeID
JOIN Products P ON OD.ProductID = P.ProductID
ORDER BY OD.OrderID, P.ProductID






(2155 rows affected)

Total execution time: 00:00:00.033

EmployeeID LastName OrderID ProductName Quantity 5 Buchanan 10248 Queso Cabrales 12 5 Buchanan 10248 Singaporean Hokkien Fried Mee 10 5 Buchanan 10248 Mozzarella di Giovanni 5 6 Suyama 10249 Tofu 9 6 Suyama 10249 Manjimup Dried Apples 40 4 Peacock 10250 Jack's New England Clam Chowder 10 4 Peacock 10250 Manjimup Dried Apples 35 4 Peacock 10250 Louisiana Fiery Hot Pepper Sauce 15 3 Leverling 10251 Gustaf's Knäckebröd 6 3 Leverling 10251 Ravioli Angelo 15 3 Leverling 10251 Louisiana Fiery Hot Pepper Sauce 20 4 Peacock 10252 Sir Rodney's Marmalade 40 4 Peacock 10252 Geitost 25 4 Peacock 10252 Camembert Pierrot 40 3 Leverling 10253 Gorgonzola Telino 20 3 Leverling 10253 Chartreuse verte 42 3 Leverling 10253 Maxilaku 40 5 Buchanan 10254 Guaraná Fantástica 15 5 Buchanan 10254 Pâté chinois 21 5 Buchanan 10254 Longlife Tofu 21 9 Dodsworth 10255 Chang 20 9 Dodsworth 10255 Pavlova 35 9 Dodsworth 10255 Inlagd Sill 25 9 Dodsworth 10255 Raclette Courdavault 30 3 Leverling 10256 Perth Pasties 15 3 Leverling 10256 Original Frankfurter grüne Soße 12 4 Peacock 10257 Schoggi Schokolade 25 4 Peacock 10257 Chartreuse verte 6 4 Peacock 10257 Original Frankfurter grüne Soße 15 1 Davolio 10258 Chang 50 1 Davolio 10258 Chef Anton's Gumbo Mix 65 1 Davolio 10258 Mascarpone Fabioli 6 4 Peacock 10259 Sir Rodney's Scones 10 4 Peacock 10259 Gravad lax 1 4 Peacock 10260 Jack's New England Clam Chowder 16 4 Peacock 10260 Ravioli Angelo 50 4 Peacock 10260 Tarte au sucre 15 4 Peacock 10260 Outback Lager 21 4 Peacock 10261 Sir Rodney's Scones 20 4 Peacock 10261 Steeleye Stout 20 8 Callahan 10262 Chef Anton's Gumbo Mix 12 8 Callahan 10262 Uncle Bob's Organic Dried Pears 15 8 Callahan 10262 Gnocchi di nonna Alice 2 9 Dodsworth 10263 Pavlova 60 9 Dodsworth 10263 Guaraná Fantástica 65 9 Dodsworth 10263 Nord-Ost Matjeshering 60 9 Dodsworth 10263 Longlife Tofu 65 6 Suyama 10264 Chang 35 6 Suyama 10264 Jack's New England Clam Chowder 25 2 Fuller 10265 Alice Mutton 30 2 Fuller 10265 Outback Lager 20 3 Leverling 10266 Queso Manchego La Pastora 12 4 Peacock 10267 Boston Crab Meat 50 4 Peacock 10267 Raclette Courdavault 70 4 Peacock 10267 Lakkalikööri 15 8 Callahan 10268 Thüringer Rostbratwurst 10 8 Callahan 10268 Mozzarella di Giovanni 4 5 Buchanan 10269 Geitost 60 5 Buchanan 10269 Mozzarella di Giovanni 20 1 Davolio 10270 Inlagd Sill 30 1 Davolio 10270 Ipoh Coffee 25 6 Suyama 10271 Geitost 24 6 Suyama 10272 Sir Rodney's Marmalade 6 6 Suyama 10272 Gorgonzola Telino 40 6 Suyama 10272 Mozzarella di Giovanni 24 3 Leverling 10273 Ikura 24 3 Leverling 10273 Gorgonzola Telino 15 3 Leverling 10273 Geitost 20 3 Leverling 10273 Boston Crab Meat 60 3 Leverling 10273 Lakkalikööri 33 6 Suyama 10274 Flotemysost 20 6 Suyama 10274 Mozzarella di Giovanni 7 1 Davolio 10275 Guaraná Fantástica 12 1 Davolio 10275 Raclette Courdavault 6 8 Callahan 10276 Ikura 15 8 Callahan 10276 Konbu 10 2 Fuller 10277 Rössle Sauerkraut 20 2 Fuller 10277 Tarte au sucre 12 8 Callahan 10278 Gula Malacca 16 8 Callahan 10278 Raclette Courdavault 15 8 Callahan 10278 Vegie-spread 8 8 Callahan 10278 Röd Kaviar 25 8 Callahan 10279 Alice Mutton 15 2 Fuller 10280 Guaraná Fantástica 12 2 Fuller 10280 Pâté chinois 20 2 Fuller 10280 Rhönbräu Klosterbier 30 4 Peacock 10281 Teatime Chocolate Biscuits 1 4 Peacock 10281 Guaraná Fantástica 6 4 Peacock 10281 Steeleye Stout 4 4 Peacock 10282 Nord-Ost Matjeshering 6 4 Peacock 10282 Ravioli Angelo 2 3 Leverling 10283 Genen Shouyu 20 3 Leverling 10283 Teatime Chocolate Biscuits 18 3 Leverling 10283 Camembert Pierrot 35 3 Leverling 10283 Mozzarella di Giovanni 3 4 Peacock 10284 Schoggi Schokolade 15 4 Peacock 10284 Gula Malacca 21 4 Peacock 10284 Camembert Pierrot 20 4 Peacock 10284 Laughing Lumberjack Lager 5 1 Davolio 10285 Chai 45 1 Davolio 10285 Boston Crab Meat 40 1 Davolio 10285 Perth Pasties 36 8 Callahan 10286 Steeleye Stout 100 8 Callahan 10286 Tarte au sucre 40 8 Callahan 10287 Pavlova 40 8 Callahan 10287 Sasquatch Ale 20 8 Callahan 10287 Spegesild 15 4 Peacock 10288 Tour

<span style="font-size: 16.000000pt; font-family: 'Cambria'">30. Customers with no orders</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">There are some customers who have never actually placed an order. Show these customers.&nbsp;</span>

In [32]:
SELECT 
    Customers_CustomerID = C.CustomerID,
    Orders_CustomerID = O.CustomerID
FROM Customers C
LEFT JOIN Orders O ON C.CustomerID = O.CustomerID
WHERE O.CustomerID IS NULL
ORDER BY Customers_CustomerID

(2 rows affected)

Total execution time: 00:00:00.005

Customers_CustomerID Orders_CustomerID FISSA NULL PARIS NULL

In [33]:
SELECT 
    CustomerID
FROM Customers
WHERE CustomerID NOT IN (SELECT CustomerID FROM Orders)

(2 rows affected)

Total execution time: 00:00:00.004

CustomerID FISSA PARIS

In [34]:
SELECT 
    CustomerID
FROM Customers
WHERE NOT EXISTS 
    (SELECT CustomerID
    FROM Orders
    WHERE Orders.CustomerID = Customers.CustomerID
    )

(2 rows affected)

Total execution time: 00:00:00.004

CustomerID FISSA PARIS

<span style="font-size: 16.000000pt; font-family: 'Cambria'">31. Customers with no orders for EmployeeID 4</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">One employee (Margaret Peacock, EmployeeID 4) has placed the most orders. However, there are some customers who've never placed an order with her. Show only those customers who have never placed an order with her.&nbsp;</span>

In [46]:
WITH CTE AS (
    SELECT 
        CustomerID
    FROM Orders
    WHERE EmployeeID = 4
)


SELECT 
    CustomerID
FROM Customers C
WHERE CustomerID NOT IN (SELECT * FROM CTE)




(16 rows affected)

Total execution time: 00:00:00.003

CustomerID CONSH DUMON FISSA FRANR GROSR LAUGB LAZYK NORTS PARIS PERIC PRINI SANTG SEVES SPECD THEBI VINET

In [58]:
SELECT 
    C.CustomerID,
    O.CustomerID
FROM Customers C
LEFT JOIN Orders O ON C.CustomerID = O.CustomerID AND O.EmployeeID = 4
WHERE O.CustomerID IS NULL


(16 rows affected)

Total execution time: 00:00:00.006

CustomerID CustomerID CONSH NULL DUMON NULL FISSA NULL FRANR NULL GROSR NULL LAUGB NULL LAZYK NULL NORTS NULL PARIS NULL PERIC NULL PRINI NULL SANTG NULL SEVES NULL SPECD NULL THEBI NULL VINET NULL

In [59]:
SELECT 
    CustomerID
FROM Customers
EXCEPT
SELECT CustomerID FROM Orders WHERE EmployeeID = 4

(16 rows affected)

Total execution time: 00:00:00.007

CustomerID CONSH DUMON FISSA FRANR GROSR LAUGB LAZYK NORTS PARIS PERIC PRINI SANTG SEVES SPECD THEBI VINET

<span style="font-size: 16.000000pt; font-family: 'Cambria'">32. High-value customers</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">We want to send all of our high-value customers a special VIP gift. We're defining high- value customers as those who've made at least 1 order with a total value (not including the discount) equal to $10,000 or more. We only want to consider orders made in the year 2016.&nbsp;</span>

In [107]:
SELECT 
    C.CustomerID,
    C.CompanyName, 
    O.OrderID,
    TotalOrderAmount = SUM(OD.Quantity * OD.UnitPrice)
FROM Orders O
JOIN OrderDetails OD ON O.OrderID = OD.OrderID
JOIN Customers C on C.CustomerID = O.CustomerID AND YEAR(O.OrderDate) = 2016
GROUP BY C.CustomerID,C.CompanyName, O.OrderID
HAVING SUM(OD.Quantity * OD.UnitPrice) >= 10000
ORDER BY TotalOrderAmount DESC 


(6 rows affected)

Total execution time: 00:00:00.013

CustomerID CompanyName OrderID TotalOrderAmount QUICK QUICK-Stop 10865 17250.00 SAVEA Save-a-lot Markets 11030 16321.90 HANAR Hanari Carnes 10981 15810.00 KOENE Königlich Essen 10817 11490.70 RATTC Rattlesnake Canyon Grocery 10889 11380.00 HUNGO Hungry Owl All-Night Grocers 10897 10835.24

<span style="font-size: 16.000000pt; font-family: 'Cambria'">33. High-value customers</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">—</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">total orders</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">The manager has changed his mind. Instead of requiring that customers have at least one individual order totaling $10,000 or more, he wants to define high-value customers differently. Now, high value customers are customers who have orders totaling $15,000 or more in 2016. How would you change the answer to the problem above?&nbsp;</span>

In [110]:
SELECT 
    C.CustomerID, 
    C.CompanyName, 
    TotalOrderAmount = SUM(OD.Quantity * OD.UnitPrice)
FROM Orders O 
JOIN OrderDetails OD ON O.OrderID = OD.OrderID
JOIN Customers C ON C.CustomerID = O.CustomerID AND YEAR(O.OrderDate) = 2016
GROUP BY C.CustomerID, C.CompanyName
HAVING SUM(OD.Quantity * OD.UnitPrice) >= 15000
ORDER BY TotalOrderAmount DESC

(9 rows affected)

Total execution time: 00:00:00.013

CustomerID CompanyName TotalOrderAmount SAVEA Save-a-lot Markets 42806.25 ERNSH Ernst Handel 42598.90 QUICK QUICK-Stop 40526.99 HANAR Hanari Carnes 24238.05 HUNGO Hungry Owl All-Night Grocers 22796.34 RATTC Rattlesnake Canyon Grocery 21725.60 KOENE Königlich Essen 20204.95 FOLKO Folk och fä HB 15973.85 WHITC White Clover Markets 15278.90

<span style="font-size: 16.000000pt; font-family: 'Cambria'">34. High-value customers</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">—</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">with discount</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Change the answer from the previous problem to use the discount when calculating high- value customers. Order by the total amount, taking into consideration the discount.&nbsp;</span>

In [129]:
SELECT 
    C.CustomerID, 
    C.CompanyName, 
    TotalWithoutDiscount = SUM(OD.Quantity * OD.UnitPrice), 
    TotalWithDiscount = SUM(OD.Quantity * OD.UnitPrice *(1 - OD.Discount))  ----- Percentage off 
FROM Orders O 
JOIN OrderDetails OD ON O.OrderID = OD.OrderID
JOIN Customers C ON C.CustomerID = O.CustomerID AND YEAR(OrderDate) = 2016
GROUP BY C.CustomerID, C.CompanyName
HAVING SUM(OD.Quantity * OD.UnitPrice *(1 - OD.Discount))  >= 15000 
ORDER BY TotalWithDiscount DESC 

(8 rows affected)

Total execution time: 00:00:00.016

CustomerID CompanyName TotalWithoutDiscount TotalWithDiscount ERNSH Ernst Handel 42598.90 41210.65002441406 QUICK QUICK-Stop 40526.99 37217.315002441406 SAVEA Save-a-lot Markets 42806.25 36310.10977935791 HANAR Hanari Carnes 24238.05 23821.199989318848 RATTC Rattlesnake Canyon Grocery 21725.60 21238.270441055298 HUNGO Hungry Owl All-Night Grocers 22796.34 20402.11993408203 KOENE Königlich Essen 20204.95 19582.773986816406 WHITC White Clover Markets 15278.90 15278.89998626709

<span style="font-size: 16.000000pt; font-family: 'Cambria'">35. Month-end orders</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">At the end of the month, salespeople are likely to try much harder to get orders, to meet their month-end quotas. Show all orders made on the last day of the month. Order by EmployeeID and OrderID&nbsp;</span>

In [137]:
SELECT 
    EmployeeID, 
    OrderID,
    OrderDate = CONVERT(varchar, OrderDate, 23)
FROM Orders 
WHERE OrderDate = EOMONTH(OrderDate)
ORDER BY EmployeeID, OrderID

(24 rows affected)

Total execution time: 00:00:00.015

EmployeeID OrderID OrderDate 1 10461 2015-02-28 1 10616 2015-07-31 2 10583 2015-06-30 2 10686 2015-09-30 2 10989 2016-03-31 2 11060 2016-04-30 3 10432 2015-01-31 3 10988 2016-03-31 3 11063 2016-04-30 4 10343 2014-10-31 4 10522 2015-04-30 4 10584 2015-06-30 4 10617 2015-07-31 4 10725 2015-10-31 4 11061 2016-04-30 4 11062 2016-04-30 5 10269 2014-07-31 6 10317 2014-09-30 7 10490 2015-03-31 8 10399 2014-12-31 8 10460 2015-02-28 8 10491 2015-03-31 8 10987 2016-03-31 9 10687 2015-09-30

<span style="font-size: 16.000000pt; font-family: 'Cambria'">36. Orders with many line items</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">The Northwind mobile app developers are testing an app that customers will use to show orders. In order to make sure that even the largest orders will show up correctly on the app, they'd like some samples of orders that have lots of individual line items.</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Show the 10 orders with the most line items, in order of total line items.&nbsp;</span>

In [153]:
SELECT TOP 10
    OrderID,
    TotalOrderDetails = COUNT(*)
FROM OrderDetails
GROUP BY OrderID
ORDER BY TotalOrderDetails DESC

(10 rows affected)

Total execution time: 00:00:00.005

OrderID TotalOrderDetails 11077 25 10657 6 10847 6 10979 6 10273 5 10294 5 10309 5 10324 5 10325 5 10337 5

<span style="font-size: 16.000000pt; font-family: 'Cambria'">37. Orders</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">—</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">random assortment</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">The Northwind mobile app developers would now like to just get a random assortment of orders for beta testing on their app. Show a random set of 2% of all orders.&nbsp;</span>

In [157]:
SELECT TOP 2 PERCENT
    OrderID
FROM Orders
ORDER BY NEWID()

(17 rows affected)

Total execution time: 00:00:00.004

OrderID 10825 10442 10735 10410 10515 10588 10726 10419 10900 10737 10608 10776 10534 10940 11043 10831 10314

<span style="font-size: 16.000000pt; font-family: 'Cambria'">38. Orders</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">—</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">accidental double-entry</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">One of the salespeople has come to you with a request. She thinks that she accidentally entered a line item twice on an order, each time with a different ProductID, but the exact same quantity. She remembers that the quantity was 60 or more. Show all the OrderIDs with line items that match this, in order of OrderID.&nbsp;</span>

In [183]:
SELECT 
    OrderID,
    Quantity
FROM OrderDetails
WHERE Quantity >= 60
GROUP BY OrderID,Quantity
HAVING COUNT(*) > 1
ORDER BY OrderID

(5 rows affected)

Total execution time: 00:00:00.006

OrderID Quantity 10263 60 10263 65 10658 70 10990 65 11030 100

<span style="font-size: 16.000000pt; font-family: 'Cambria'">39. Orders</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">—</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">accidental double-entry details</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Based on the previous question, we now want to show details of the order, for orders that match the above criteria.&nbsp;</span>

In [196]:
WITH CTE AS(
    SELECT 
    OrderID
FROM OrderDetails
WHERE Quantity >= 60
GROUP BY OrderID,Quantity
HAVING COUNT(*) > 1
)



SELECT 
    *
FROM OrderDetails
WHERE OrderID IN (SELECT * FROM CTE)
ORDER BY OrderID, Quantity
 Looks  

(16 rows affected)

Total execution time: 00:00:00.015

OrderID ProductID UnitPrice Quantity Discount 10263 16 13.90 60 0.25 10263 30 20.70 60 0.25 10263 24 3.60 65 0 10263 74 8.00 65 0.25 10658 60 34.00 55 0.05 10658 21 10.00 60 0 10658 40 18.40 70 0.05 10658 77 13.00 70 0.05 10990 34 14.00 60 0.15 10990 21 10.00 65 0 10990 55 24.00 65 0.15 10990 61 28.50 66 0.15 11030 29 123.79 60 0.25 11030 5 21.35 70 0 11030 2 19.00 100 0.25 11030 59 55.00 100 0.25

<span style="font-size: 16.000000pt; font-family: 'Cambria'">40. Orders</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">—</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">accidental double-entry details, derived table</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Here's another way of getting the same results as in the previous problem, using a derived table instead of a CTE. However, there's a bug in this SQL. It returns 20 rows instead of 16. Correct the SQL.&nbsp;</span> 

<span style="font-family: TimesNewRomanPSMT; font-size: 12pt; color: var(--vscode-foreground);">Problem SQL:&nbsp;</span>   

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'"><br></span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'"><br></span>

In [199]:
SELECT 
    *
FROM OrderDetails  JOIN (
    SELECT DISTINCT
        OrderID
    FROM OrderDetails
    WHERE Quantity >=60
    GROUP BY ORDERID, Quantity
    HAVING COUNT(*) > 1
) PotentialProblemOrders 
ON PotentialProblemOrders.OrderID = OrderDetails.OrderID
ORDER BY OrderDetails.OrderID,ProductID

(16 rows affected)

Total execution time: 00:00:00.017

OrderID ProductID UnitPrice Quantity Discount OrderID 10263 16 13.90 60 0.25 10263 10263 24 3.60 65 0 10263 10263 30 20.70 60 0.25 10263 10263 74 8.00 65 0.25 10263 10658 21 10.00 60 0 10658 10658 40 18.40 70 0.05 10658 10658 60 34.00 55 0.05 10658 10658 77 13.00 70 0.05 10658 10990 21 10.00 65 0 10990 10990 34 14.00 60 0.15 10990 10990 55 24.00 65 0.15 10990 10990 61 28.50 66 0.15 10990 11030 2 19.00 100 0.25 11030 11030 5 21.35 70 0 11030 11030 29 123.79 60 0.25 11030 11030 59 55.00 100 0.25 11030

<span style="font-size: 16.000000pt; font-family: 'Cambria'">41. Late orders</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Some customers are complaining about their orders arriving late. Which orders are late? Sort the results by OrderID.&nbsp;</span>

In [205]:
SELECT 
    OrderID,
    OrderDate = CONVERT(date, OrderDate),
    RequiredDate = CONVERT(date, RequiredDate),
    ShippedDate = CONVERT(date, ShippedDate)
FROM Orders
WHERE ShippedDate >= RequiredDate
ORDER BY OrderID 

(39 rows affected)

Total execution time: 00:00:00.004

OrderID OrderDate RequiredDate ShippedDate 10264 2014-07-24 2014-08-21 2014-08-23 10271 2014-08-01 2014-08-29 2014-08-30 10280 2014-08-14 2014-09-11 2014-09-12 10302 2014-09-10 2014-10-08 2014-10-09 10309 2014-09-19 2014-10-17 2014-10-23 10380 2014-12-12 2015-01-09 2015-01-16 10423 2015-01-23 2015-02-06 2015-02-24 10427 2015-01-27 2015-02-24 2015-03-03 10433 2015-02-03 2015-03-03 2015-03-04 10451 2015-02-19 2015-03-05 2015-03-12 10483 2015-03-24 2015-04-21 2015-04-25 10515 2015-04-23 2015-05-07 2015-05-23 10523 2015-05-01 2015-05-29 2015-05-30 10545 2015-05-22 2015-06-19 2015-06-26 10578 2015-06-24 2015-07-22 2015-07-25 10593 2015-07-09 2015-08-06 2015-08-13 10596 2015-07-11 2015-08-08 2015-08-12 10660 2015-09-08 2015-10-06 2015-10-15 10663 2015-09-10 2015-09-24 2015-10-03 10687 2015-09-30 2015-10-28 2015-10-30 10705 2015-10-15 2015-11-12 2015-11-18 10709 2015-10-17 2015-11-14 2015-11-20 10726 2015-11-03 2015-11-17 2015-12-05 10727 2015-11-03 2015-12-01 2015-12-05 10749 2015-11-20 2015-12-18 2015-12-19 10777 2015-12-15 2015-12-29 2016-01-21 10779 2015-12-16 2016-01-13 2016-01-14 10788 2015-12-22 2016-01-19 2016-01-19 10807 2015-12-31 2016-01-28 2016-01-30 10816 2016-01-06 2016-02-03 2016-02-04 10827 2016-01-12 2016-01-26 2016-02-06 10828 2016-01-13 2016-01-27 2016-02-04 10847 2016-01-22 2016-02-05 2016-02-10 10924 2016-03-04 2016-04-01 2016-04-08 10927 2016-03-05 2016-04-02 2016-04-08 10960 2016-03-19 2016-04-02 2016-04-08 10970 2016-03-24 2016-04-07 2016-04-24 10978 2016-03-26 2016-04-23 2016-04-23 10998 2016-04-03 2016-04-17 2016-04-17

<span style="font-size: 16.000000pt; font-family: 'Cambria'">42. Late orders</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">—</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">which employees?</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Some salespeople have more orders arriving late than others. Maybe they're not following up on the order process, and need more training. Which salespeople have the most orders arriving late?&nbsp;</span>

In [216]:
WITH LateOrders AS (SELECT 
    *
FROM Orders
WHERE ShippedDate >= RequiredDate
) 


SELECT 
    E.EmployeeID,
    E.LastName,
    TotalLateOrders = COUNT(*)
FROM Employees E
JOIN LateOrders L ON E.EmployeeID = L.EmployeeID
GROUP BY E.EmployeeID,E.LastName
ORDER BY TotalLateOrders DESC 

(8 rows affected)

Total execution time: 00:00:00.008

EmployeeID LastName TotalLateOrders 4 Peacock 10 3 Leverling 5 8 Callahan 5 9 Dodsworth 5 7 King 4 2 Fuller 4 1 Davolio 3 6 Suyama 3

In [5]:
SELECT 
OrderID,
AllOrders = COUNT(*)
FROM Orders
GROUP BY OrderID 
ORDER BY AllOrders DESC

(830 rows affected)

Total execution time: 00:00:00.011

OrderID AllOrders 10248 1 10249 1 10250 1 10251 1 10252 1 10253 1 10254 1 10255 1 10256 1 10257 1 10258 1 10259 1 10260 1 10261 1 10262 1 10263 1 10264 1 10265 1 10266 1 10267 1 10268 1 10269 1 10270 1 10271 1 10272 1 10273 1 10274 1 10275 1 10276 1 10277 1 10278 1 10279 1 10280 1 10281 1 10282 1 10283 1 10284 1 10285 1 10286 1 10287 1 10288 1 10289 1 10290 1 10291 1 10292 1 10293 1 10294 1 10295 1 10296 1 10297 1 10298 1 10299 1 10300 1 10301 1 10302 1 10303 1 10304 1 10305 1 10306 1 10307 1 10308 1 10309 1 10310 1 10311 1 10312 1 10313 1 10314 1 10315 1 10316 1 10317 1 10318 1 10319 1 10320 1 10321 1 10322 1 10323 1 10324 1 10325 1 10326 1 10327 1 10328 1 10329 1 10330 1 10331 1 10332 1 10333 1 10334 1 10335 1 10336 1 10337 1 10338 1 10339 1 10340 1 10341 1 10342 1 10343 1 10344 1 10345 1 10346 1 10347 1 10348 1 10349 1 10350 1 10351 1 10352 1 10353 1 10354 1 10355 1 10356 1 10357 1 10358 1 10359 1 10360 1 10361 1 10362 1 10363 1 10364 1 10365 1 10366 1 10367 1 10368 1 10369 1 10370 1 10371 1 10372 1 10373 1 10374 1 10375 1 10376 1 10377 1 10378 1 10379 1 10380 1 10381 1 10382 1 10383 1 10384 1 10385 1 10386 1 10387 1 10388 1 10389 1 10390 1 10391 1 10392 1 10393 1 10394 1 10395 1 10396 1 10397 1 10398 1 10399 1 10400 1 10401 1 10402 1 10403 1 10404 1 10405 1 10406 1 10407 1 10408 1 10409 1 10410 1 10411 1 10412 1 10413 1 10414 1 10415 1 10416 1 10417 1 10418 1 10419 1 10420 1 10421 1 10422 1 10423 1 10424 1 10425 1 10426 1 10427 1 10428 1 10429 1 10430 1 10431 1 10432 1 10433 1 10434 1 10435 1 10436 1 10437 1 10438 1 10439 1 10440 1 10441 1 10442 1 10443 1 10444 1 10445 1 10446 1 10447 1 10448 1 10449 1 10450 1 10451 1 10452 1 10453 1 10454 1 10455 1 10456 1 10457 1 10458 1 10459 1 10460 1 10461 1 10462 1 10463 1 10464 1 10465 1 10466 1 10467 1 10468 1 10469 1 10470 1 10471 1 10472 1 10473 1 10474 1 10475 1 10476 1 10477 1 10478 1 10479 1 10480 1 10481 1 10482 1 10483 1 10484 1 10485 1 10486 1 10487 1 10488 1 10489 1 10490 1 10491 1 10492 1 10493 1 10494 1 10495 1 10496 1 10497 1 10498 1 10499 1 10500 1 10501 1 10502 1 10503 1 10504 1 10505 1 10506 1 10507 1 10508 1 10509 1 10510 1 10511 1 10512 1 10513 1 10514 1 10515 1 10516 1 10517 1 10518 1 10519 1 10520 1 10521 1 10522 1 10523 1 10524 1 10525 1 10526 1 10527 1 10528 1 10529 1 10530 1 10531 1 10532 1 10533 1 10534 1 10535 1 10536 1 10537 1 10538 1 10539 1 10540 1 10541 1 10542 1 10543 1 10544 1 10545 1 10546 1 10547 1 10548 1 10549 1 10550 1 10551 1 10552 1 10553 1 10554 1 10555 1 10556 1 10557 1 10558 1 10559 1 10560 1 10561 1 10562 1 10563 1 10564 1 10565 1 10566 1 10567 1 10568 1 10569 1 10570 1 10571 1 10572 1 10573 1 10574 1 10575 1 10576 1 10577 1 10578 1 10579 1 10580 1 10581 1 10582 1 10583 1 10584 1 10585 1 10586 1 10587 1 10588 1 10589 1 10590 1 10591 1 10592 1 10593 1 10594 1 10595 1 10596 1 10597 1 10598 1 10599 1 10600 1 10601 1 10602 1 10603 1 10604 1 10605 1 10606 1 10607 1 10608 1 10609 1 10610 1 10611 1 10612 1 10613 1 10614 1 10615 1 10616 1 10617 1 10618 1 10619 1 10620 1 10621 1 10622 1 10623 1 10624 1 10625 1 10626 1 10627 1 10628 1 10629 1 10630 1 10631 1 10632 1 10633 1 10634 1 10635 1 10636 1 10637 1 10638 1 10639 1 10640 1 10641 1 10642 1 10643 1 10644 1 10645 1 10646 1 10647 1 10648 1 10649 1 10650 1 10651 1 10652 1 10653 1 10654 1 10655 1 10656 1 10657 1 10658 1 10659 1 10660 1 10661 1 10662 1 10663 1 10664 1 10665 1 10666 1 10667 1 10668 1 10669 1 10670 1 10671 1 10672 1 10673 1 10674 1 10675 1 10676 1 10677 1 10678 1 10679 1 10680 1 10681 1 10682 1 10683 1 10684 1 10685 1 10686 1 10687 1 10688 1 10689 1 10690 1 10691 1 10692 1 10693 1 10694 1 10695 1 10696 1 10697 1 10698 1 10699 1 10700 1 10701 1 10702 1 10703 1 10704 1 10705 1 10706 1 10707 1 10708 1 10709 1 10710 1 10711 1 10712 1 10713 1 10714 1 10715 1 10716 1 10717 1 10718 1 10719 1 10720 1 10721 1 10722 1 10723 1 10724 1 10725 1 10726 1 10727 1 10728 1 10729 1 10730 1 10731 1 10732 1 10733 1 10734 1 10735 1 10736 1 10737 1 10738 1 10739 1 10740 1 10741 1 10742 1 10743 1 10744 1 10745 

<span style="font-size: 16.000000pt; font-family: 'Cambria'">43. Late orders vs. total orders</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Andrew, the VP of sales, has been doing some more thinking some more about the problem of late orders. He realizes that just looking at the number of orders arriving late for each salesperson isn't a good idea. It needs to be compared to the </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPS'; font-style: italic">total </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">number of orders per salesperson. We want results like the following:&nbsp;</span>

In [22]:
WITH LateOrders AS (
        SELECT 
        EmployeeID,
        LateOrders = count(*)
        FROM Orders 
        WHERE ShippedDate >= RequiredDate
        GROUP BY EmployeeID
), AllOrders AS (
    SELECT 
    EmployeeID,
    AllOrders = COUNT(*)
    FROM Orders
    GROUP BY EmployeeID
)

SELECT 
    E.EmployeeID,
    E.LastName,
    A.AllOrders,
    L.LateOrders
FROM Employees E 
JOIN LateOrders L on E.EmployeeID = L.EmployeeID
JOIN AllOrders A on E.EmployeeID = A.EmployeeID
ORDER BY E.EmployeeID

(8 rows affected)

Total execution time: 00:00:00.013

EmployeeID LastName AllOrders LateOrders 1 Davolio 123 3 2 Fuller 96 4 3 Leverling 127 5 4 Peacock 156 10 6 Suyama 67 3 7 King 72 4 8 Callahan 104 5 9 Dodsworth 43 5

<span style="font-size: 16.000000pt; font-family: 'Cambria'">44. Late orders vs. total orders</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">—</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">missing employee</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">There's an employee missing in the answer from the problem above. Fix the SQL to show all employees who have taken orders.&nbsp;</span>

In [23]:
WITH LateOrders AS (
        SELECT 
        EmployeeID,
        LateOrders = count(*)
        FROM Orders 
        WHERE ShippedDate >= RequiredDate
        GROUP BY EmployeeID
), AllOrders AS (
    SELECT 
    EmployeeID,
    AllOrders = COUNT(*)
    FROM Orders
    GROUP BY EmployeeID
)

SELECT 
    E.EmployeeID,
    E.LastName,
    A.AllOrders,
    L.LateOrders
FROM Employees E 
FULL JOIN LateOrders L on E.EmployeeID = L.EmployeeID
JOIN AllOrders A on E.EmployeeID = A.EmployeeID
ORDER BY E.EmployeeID

(9 rows affected)

Total execution time: 00:00:00.015

EmployeeID LastName AllOrders LateOrders 1 Davolio 123 3 2 Fuller 96 4 3 Leverling 127 5 4 Peacock 156 10 5 Buchanan 42 NULL 6 Suyama 67 3 7 King 72 4 8 Callahan 104 5 9 Dodsworth 43 5

<span style="font-size: 16.000000pt; font-family: 'Cambria'">45. Late orders vs. total orders</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">—</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">fix null</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Continuing on the answer for the previous problem, let's fix the results for row 5 - Buchanan. He should have a 0 instead of a Null in LateOrders.&nbsp;</span>

In [29]:
WITH LateOrders AS (
        SELECT 
        EmployeeID,
        LateOrders = count(*)
        FROM Orders 
        WHERE ShippedDate >= RequiredDate
        GROUP BY EmployeeID
), AllOrders AS (
    SELECT 
    EmployeeID,
    AllOrders = COUNT(*)
    FROM Orders
    GROUP BY EmployeeID
)

SELECT 
    E.EmployeeID,
    E.LastName,
    AllOrders = ISNULL(A.AllOrders,0),
    LateOrders = ISNULL(L.LateOrders,0)
FROM Employees E 
FULL JOIN LateOrders L on E.EmployeeID = L.EmployeeID
JOIN AllOrders A on E.EmployeeID = A.EmployeeID
ORDER BY E.EmployeeID

(9 rows affected)

Total execution time: 00:00:00.021

EmployeeID LastName AllOrders LateOrders 1 Davolio 123 3 2 Fuller 96 4 3 Leverling 127 5 4 Peacock 156 10 5 Buchanan 42 0 6 Suyama 67 3 7 King 72 4 8 Callahan 104 5 9 Dodsworth 43 5

In [30]:
WITH LateOrders AS (
        SELECT 
        EmployeeID,
        LateOrders = count(*)
        FROM Orders 
        WHERE ShippedDate >= RequiredDate
        GROUP BY EmployeeID
), AllOrders AS (
    SELECT 
    EmployeeID,
    AllOrders = COUNT(*)
    FROM Orders
    GROUP BY EmployeeID
)

SELECT 
    E.EmployeeID,
    E.LastName,
    AllOrders = COALESCE(A.AllOrders,0),
    LateOrders = CASE   
                    WHEN L.LateOrders IS NULL THEN 0
                    ELSE L.LateOrders
                END
FROM Employees E 
FULL JOIN LateOrders L on E.EmployeeID = L.EmployeeID
JOIN AllOrders A on E.EmployeeID = A.EmployeeID
ORDER BY E.EmployeeID

(9 rows affected)

Total execution time: 00:00:00.013

EmployeeID LastName AllOrders LateOrders 1 Davolio 123 3 2 Fuller 96 4 3 Leverling 127 5 4 Peacock 156 10 5 Buchanan 42 0 6 Suyama 67 3 7 King 72 4 8 Callahan 104 5 9 Dodsworth 43 5

<span style="font-size: 16.000000pt; font-family: 'Cambria'">46. Late orders vs. total orders</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">—</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">percentage</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Now we want to get the percentage of late orders over total orders.&nbsp;</span>

In [41]:
WITH LateOrders AS (
        SELECT 
        EmployeeID,
        LateOrders = count(*)
        FROM Orders 
        WHERE ShippedDate >= RequiredDate
        GROUP BY EmployeeID
), AllOrders AS (
    SELECT 
    EmployeeID,
    AllOrders = COUNT(*)
    FROM Orders
    GROUP BY EmployeeID
)

SELECT 
    E.EmployeeID,
    E.LastName,
    AllOrders = ISNULL(A.AllOrders,0),
    LateOrders = ISNULL(L.LateOrders,0),
    PercentLateOrders = ISNULL(CAST(LateOrders AS float)/CAST(AllOrders AS float),0)
FROM Employees E 
FULL JOIN LateOrders L on E.EmployeeID = L.EmployeeID
JOIN AllOrders A on E.EmployeeID = A.EmployeeID
ORDER BY E.EmployeeID

(9 rows affected)

Total execution time: 00:00:00.015

EmployeeID LastName AllOrders LateOrders PercentLateOrders 1 Davolio 123 3 0.024390243902439025 2 Fuller 96 4 0.041666666666666664 3 Leverling 127 5 0.03937007874015748 4 Peacock 156 10 0.0641025641025641 5 Buchanan 42 0 0 6 Suyama 67 3 0.04477611940298507 7 King 72 4 0.05555555555555555 8 Callahan 104 5 0.04807692307692308 9 Dodsworth 43 5 0.11627906976744186

In [43]:
WITH LateOrders AS (
        SELECT 
        EmployeeID,
        LateOrders = count(*)
        FROM Orders 
        WHERE ShippedDate >= RequiredDate
        GROUP BY EmployeeID
), AllOrders AS (
    SELECT 
    EmployeeID,
    AllOrders = COUNT(*)
    FROM Orders
    GROUP BY EmployeeID
)

SELECT 
    E.EmployeeID,
    E.LastName,
    AllOrders = ISNULL(A.AllOrders,0),
    LateOrders = ISNULL(L.LateOrders,0),
    PercentLateOrders = ISNULL(LateOrders,0)*1.00/AllOrders
FROM Employees E 
FULL JOIN LateOrders L on E.EmployeeID = L.EmployeeID
JOIN AllOrders A on E.EmployeeID = A.EmployeeID
ORDER BY E.EmployeeID

(9 rows affected)

Total execution time: 00:00:00.018

EmployeeID LastName AllOrders LateOrders PercentLateOrders 1 Davolio 123 3 0.0243902439024 2 Fuller 96 4 0.0416666666666 3 Leverling 127 5 0.0393700787401 4 Peacock 156 10 0.0641025641025 5 Buchanan 42 0 0.0000000000000 6 Suyama 67 3 0.0447761194029 7 King 72 4 0.0555555555555 8 Callahan 104 5 0.0480769230769 9 Dodsworth 43 5 0.1162790697674

<span style="font-size: 16.000000pt; font-family: 'Cambria'">47. Late orders vs. total orders</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">—</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">fix decimal</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">So now for the PercentageLateOrders, we get a decimal value like we should. But to make the output easier to read, let's cut the PercentLateOrders off at 2 digits to the right of the decimal point.&nbsp;</span>

In [45]:
WITH LateOrders AS (
        SELECT 
        EmployeeID,
        LateOrders = count(*)
        FROM Orders 
        WHERE ShippedDate >= RequiredDate
        GROUP BY EmployeeID
), AllOrders AS (
    SELECT 
    EmployeeID,
    AllOrders = COUNT(*)
    FROM Orders
    GROUP BY EmployeeID
)

SELECT 
    E.EmployeeID,
    E.LastName,
    AllOrders = ISNULL(A.AllOrders,0),
    LateOrders = ISNULL(L.LateOrders,0),
    PercentLateOrders = CAST(ISNULL(LateOrders,0)*1.00/AllOrders AS decimal(2,2))
FROM Employees E 
FULL JOIN LateOrders L on E.EmployeeID = L.EmployeeID
JOIN AllOrders A on E.EmployeeID = A.EmployeeID
ORDER BY E.EmployeeID

(9 rows affected)

Total execution time: 00:00:00.017

EmployeeID LastName AllOrders LateOrders PercentLateOrders 1 Davolio 123 3 0.02 2 Fuller 96 4 0.04 3 Leverling 127 5 0.04 4 Peacock 156 10 0.06 5 Buchanan 42 0 0.00 6 Suyama 67 3 0.04 7 King 72 4 0.06 8 Callahan 104 5 0.05 9 Dodsworth 43 5 0.12

In [46]:
WITH LateOrders AS (
        SELECT 
        EmployeeID,
        LateOrders = count(*)
        FROM Orders 
        WHERE ShippedDate >= RequiredDate
        GROUP BY EmployeeID
), AllOrders AS (
    SELECT 
    EmployeeID,
    AllOrders = COUNT(*)
    FROM Orders
    GROUP BY EmployeeID
)

SELECT 
    E.EmployeeID,
    E.LastName,
    AllOrders = ISNULL(A.AllOrders,0),
    LateOrders = ISNULL(L.LateOrders,0),
    PercentLateOrders = CONVERT(decimal(2,2),ISNULL(LateOrders,0)*1.00/AllOrders)
FROM Employees E 
FULL JOIN LateOrders L on E.EmployeeID = L.EmployeeID
JOIN AllOrders A on E.EmployeeID = A.EmployeeID
ORDER BY E.EmployeeID

(9 rows affected)

Total execution time: 00:00:00.016

EmployeeID LastName AllOrders LateOrders PercentLateOrders 1 Davolio 123 3 0.02 2 Fuller 96 4 0.04 3 Leverling 127 5 0.04 4 Peacock 156 10 0.06 5 Buchanan 42 0 0.00 6 Suyama 67 3 0.04 7 King 72 4 0.06 8 Callahan 104 5 0.05 9 Dodsworth 43 5 0.12

<span style="font-size: 16.000000pt; font-family: 'Cambria'">48. Customer grouping</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Andrew Fuller, the VP of sales at Northwind, would like to do a sales campaign for existing customers. He'd like to categorize customers into groups, based on how much they ordered in 2016. Then, depending on which group the customer is in, he will target the customer with different sales materials.</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">The customer grouping categories are 0 to 1,000, 1,000 to 5,000, 5,000 to 10,000, and over 10,000. </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">So, if the total dollar amount of the customer’s purchases in that year </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">were between </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">0 to 1,000, they would be in the “Low” group. A customer with purchase from 1,000 to 5,000 would be in the “Medium” group, and so on.</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">A good starting point for this query is the answer from the problem </span> <span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">“</span><span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">High-value customers</span><span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">—</span><span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">total orders</span><span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">”</span><span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">. Also, we only want to show customers who have ordered in 2016.</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Order the results by CustomerID.&nbsp;</span> 

<span style="font-size: 16.000000pt; font-family: 'Cambria'">49. Customer grouping</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">—</span><span style="font-size: 16.000000pt; font-family: 'Cambria'">fix null&nbsp;</span> 

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">There's a problem with the answer to the previous question. The CustomerGroup value for one of the rows is null.</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Fix the SQL so that there are no nulls in the CustomerGroup field.&nbsp;</span>

In [99]:
WITH TotalOrderAmount AS (
SELECT 
    O.CustomerID,
    C.CompanyName,
    TotalOrderAmount = SUM(Quantity*UnitPrice)
FROM OrderDetails OD 
JOIN Orders O ON OD.OrderID = O.OrderID AND YEAR(O.OrderDate) = 2016
JOIN Customers C ON O.CustomerID = C.CustomerID
GROUP BY O.CustomerID, C.CompanyName
)


SELECT 
    *,
   CustomerGroup = CASE 
        WHEN TotalOrderAmount >= 0 AND TotalOrderAmount <= 1000 THEN 'Low'
        WHEN TotalOrderAmount >= 1000 AND TotalOrderAmount <=5000 THEN 'Medium'
        WHEN TotalOrderAmount >=5000 AND TotalOrderAmount <=10000 THEN 'High'
        ELSE 'Very High'
    END
FROM TotalOrderAmount





(81 rows affected)

Total execution time: 00:00:00.016

CustomerID CompanyName TotalOrderAmount CustomerGroup ALFKI Alfreds Futterkiste 2302.20 Medium ANATR Ana Trujillo Emparedados y helados 514.40 Low ANTON Antonio Moreno Taquería 660.00 Low AROUT Around the Horn 5838.50 High BERGS Berglunds snabbköp 8110.55 High BLAUS Blauer See Delikatessen 2160.00 Medium BLONP Blondesddsl père et fils 730.00 Low BOLID Bólido Comidas preparadas 280.00 Low BONAP Bon app' 7185.90 High BOTTM Bottom-Dollar Markets 12227.40 Very High BSBEV B's Beverages 2431.00 Medium CACTU Cactus Comidas para llevar 1576.80 Medium CHOPS Chop-suey Chinese 4429.40 Medium COMMI Comércio Mineiro 513.75 Low CONSH Consolidated Holdings 931.50 Low DRACD Drachenblut Delikatessen 2809.61 Medium DUMON Du monde entier 860.10 Low EASTC Eastern Connection 9569.31 High ERNSH Ernst Handel 42598.90 Very High FOLKO Folk och fä HB 15973.85 Very High FRANK Frankenversand 5587.00 High FRANR France restauration 2252.06 Medium FRANS Franchi S.p.A. 1296.00 Medium FURIB Furia Bacalhau e Frutos do Mar 68.00 Low GALED Galería del gastrónomo 207.50 Low GODOS Godos Cocina Típica 7064.05 High GOURL Gourmet Lanchonetes 497.00 Low GREAL Great Lakes Food Market 10562.58 Very High HANAR Hanari Carnes 24238.05 Very High HILAA HILARION-Abastos 6132.30 High HUNGO Hungry Owl All-Night Grocers 22796.34 Very High ISLAT Island Trading 2684.60 Medium KOENE Königlich Essen 20204.95 Very High LACOR La corne d'abondance 1992.05 Medium LAMAI La maison d'Asie 1549.95 Medium LAUGB Laughing Bacchus Wine Cellars 187.00 Low LEHMS Lehmanns Marktstand 3342.85 Medium LETSS Let's Stop N Shop 1450.60 Medium LILAS LILA-Supermercado 5994.06 High LINOD LINO-Delicateses 10085.60 Very High LONEP Lonesome Pine Restaurant 1709.40 Medium MAGAA Magazzini Alimentari Riuniti 1693.00 Medium MAISD Maison Dewey 5000.20 High MORGK Morgenstern Gesundkost 245.00 Low NORTS North/South 45.00 Low OCEAN Océano Atlántico Ltda. 3031.00 Medium OLDWO Old World Delicatessen 5337.65 High OTTIK Ottilies Käseladen 3012.70 Medium PERIC Pericles Comidas clásicas 1496.00 Medium PICCO Piccolo und mehr 4393.75 Medium PRINI Princesa Isabel Vinhos 2633.90 Medium QUEDE Que Delícia 1353.60 Medium QUEEN Queen Cozinha 7007.65 High QUICK QUICK-Stop 40526.99 Very High RANCH Rancho grande 1694.70 Medium RATTC Rattlesnake Canyon Grocery 21725.60 Very High REGGC Reggiani Caseifici 4263.00 Medium RICAR Ricardo Adocicados 7312.00 High RICSU Richter Supermarkt 5497.90 High ROMEY Romero y tomillo 726.89 Low SANTG Santé Gourmet 3976.75 Medium SAVEA Save-a-lot Markets 42806.25 Very High SEVES Seven Seas Imports 1630.00 Medium SIMOB Simons bistro 244.30 Low SPECD Spécialités du monde 2371.00 Medium SPLIR Split Rail Beer & Ale 1117.00 Medium SUPRD Suprêmes délices 11862.50 Very High THEBI The Big Cheese 69.60 Low THECR The Cracker Box 326.00 Low TOMSP Toms Spezialitäten 910.40 Low TORTU Tortuga Restaurante 1874.50 Medium TRADH Tradição Hipermercados 4401.62 Medium TRAIH Trail's Head Gourmet Provisioners 237.90 Low VAFFE Vaffeljernet 4333.50 Medium VICTE Victuailles en stock 3022.00 Medium WANDK Die Wandernde Kuh 1564.00 Medium WARTH Wartian Herkku 300.00 Low WELLI Wellington Importadora 1245.00 Medium WHITC White Clover Markets 15278.90 Very High WILMK Wilman Kala 1987.00 Medium WOLZA Wolski Zajazd 1865.10 Medium

<span style="font-size: 16.000000pt; font-family: 'Cambria'">50. Customer grouping with percentage</span>

<span style="font-size: 12.000000pt; font-family: 'TimesNewRomanPSMT'">Based on the above query, show all the defined CustomerGroups, and the percentage in each. Sort by the total in each group, in descending order.&nbsp;</span>

In [126]:
WITH TotalOrderAmount AS (
SELECT 
    O.CustomerID,
    C.CompanyName,
    TotalOrderAmount = SUM(Quantity*UnitPrice)
FROM OrderDetails OD 
JOIN Orders O ON OD.OrderID = O.OrderID AND YEAR(O.OrderDate) = 2016
JOIN Customers C ON O.CustomerID = C.CustomerID
GROUP BY O.CustomerID, C.CompanyName
),
CustomerGrouping AS (SELECT 
   CustomerGroup = CASE 
        WHEN TotalOrderAmount >= 0 AND TotalOrderAmount <= 1000 THEN 'Low'
        WHEN TotalOrderAmount >= 1000 AND TotalOrderAmount <=5000 THEN 'Medium'
        WHEN TotalOrderAmount >=5000 AND TotalOrderAmount <=10000 THEN 'High'
        ELSE 'Very High'
    END
FROM TotalOrderAmount)

SELECT 
    CustomerGroup,
    TotalInGRoup = COUNT(*),
    PercentageInGroup = COUNT(*) * 1.00 / (SELECT COUNT(*) FROM CustomerGrouping)
FROM CustomerGrouping
GROUP BY CustomerGroup
ORDER BY TotalInGRoup DESC 


(4 rows affected)

Total execution time: 00:00:00.057

CustomerGroup TotalInGRoup PercentageInGroup Medium 35 0.4320987654320 Low 20 0.2469135802469 High 13 0.1604938271604 Very High 13 0.1604938271604